src/devices/sound/hc55516.cpp

```diff
@@ -21,242 +31,563 @@
+	return (((uint64_t)m_samples_generated * clock() * 2 / SAMPLE_RATE) & 0x01)?true:false;
+}
 
-	/* create the stream */
-	m_channel = stream_alloc(0, 1, SAMPLE_RATE);
+void cvsd_device::digit_w(int digit)
+{
+	m_stream->update();
+	m_buffered_bit = digit ? true : false;
+}
 
-	save_item(NAME(m_last_clock_state));
-	save_item(NAME(m_digit));
-	save_item(NAME(m_new_digit));
-	save_item(NAME(m_shiftreg));
-	save_item(NAME(m_curr_sample));
-	save_item(NAME(m_next_sample));
-	save_item(NAME(m_update_count));
-	save_item(NAME(m_filter));
-	save_item(NAME(m_integrator));
+void cvsd_device::clock_w(int state)
+{
+	/* update the output buffer first */
+	m_stream->update();
+	bool clock_state = state ? true : false;
+
+	/* only makes sense for setups with a software driven clock */
+	assert(!is_external_oscillator());
+
+	/* speech clock changing? */
+	if (is_clock_changed(clock_state))
+	{
+		/* clear the update count */
+		m_samples_generated = 0;
+		process_bit(m_buffered_bit, clock_state);
+	}
+
+	/* update the clock */
+	m_last_clock_state = clock_state;
 }
 
-inline int hc55516_device::is_external_oscillator()
+int cvsd_device::clock_state_r()
 {
-	return clock() != 0;
+	/* only makes sense for setups with an external oscillator */
+	assert(is_external_oscillator());
+
+	m_stream->update();
+
+	return current_clock_state();
+}
+
+void cvsd_device::process_bit(bool bit, bool clock_state)
+{
+	// stub
 }
 
+//-------------------------------------------------
+//  sound_stream_update - handle a stream update
+//-------------------------------------------------
 
-inline int hc55516_device::is_active_clock_transition(int clock_state)
+void cvsd_device::sound_stream_update(sound_stream &stream, std::vector<read_stream_view> const &inputs, std::vector<write_stream_view> &outputs)
 {
-	return (( m_active_clock_hi && !m_last_clock_state &&  clock_state) ||
-			(!m_active_clock_hi &&  m_last_clock_state && !clock_state));
+	// Stub, just return silence
+	auto &buffer = outputs[0];
+
+	m_samples_generated += buffer.samples();
+	if (m_samples_generated >= SAMPLE_RATE)
+		m_samples_generated -= SAMPLE_RATE;
+	for (int i = 0; i < buffer.samples(); i++)
+		buffer.put(i, stream_buffer::sample_t(0.0));
 }
 
 
-inline int hc55516_device::current_clock_state()
+//#########################################
+//                 HC55516
+//#########################################
+DEFINE_DEVICE_TYPE(HC55516, hc55516_device, "hc55516", "HC-55516")
+
+hc55516_device::hc55516_device(const machine_config &mconfig, const char *tag, device_t *owner, uint32_t clock)
+	: cvsd_device(mconfig, HC55516, tag, owner, clock)
+	, m_agc_push_cb(*this)
+	, m_fzq_pull_cb(*this)
+	, m_sylmask(0xfc0)
+	, m_sylshift(6)
+	, m_syladd(0xfc1)
+	, m_intshift(4)
+	, m_sylfilter(0)
+	, m_intfilter(0)
+	, m_agc(true)
+	, m_buffered_fzq(true)
 {
-	return ((uint64_t)m_update_count * clock() * 2 / SAMPLE_RATE) & 0x01;
 }
```

Comment:
Use constructor delegation to avoid repeating yourself.

Reply:
This is reduced as there are less parameters initialized in both hc55516 and hc55532 now that several are constants.

Reply:
constructor delegation is now done.